File: src/main/scala/uk/co/developmentanddinosaurs/gatling/lambda/LambdaDsl.scala
package uk.co.developmentanddinosaurs.gatling.lambda

import scala.language.implicitConversions
import io.gatling.core.action.builder.ActionBuilder
import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression
import uk.co.developmentanddinosaurs.gatling.lambda.protocol.{LambdaProtocol, LambdaProtocolBuilder}
import uk.co.developmentanddinosaurs.gatling.lambda.request.LambdaInvokeBuilder

class LambdaDsl {

  def lambda(function: Expression[String]): LambdaInvokeBuilder =
    LambdaInvokeBuilder(function)

  def lambda(
      implicit configuration: GatlingConfiguration
  ): LambdaProtocolBuilder = LambdaProtocolBuilder()

  implicit def invokeBuilderToActionBuilder(
      builder: LambdaInvokeBuilder
  ): ActionBuilder = builder.build

  implicit def protocolBuilderToProtocol(
      builder: LambdaProtocolBuilder
  ): LambdaProtocol = builder.build
}

File: src/main/java/uk/co/developmentanddinosaurs/gatling/lambda/javaapi/LambdaDsl.java
package uk.co.developmentanddinosaurs.gatling.lambda.javaapi;

import io.gatling.core.Predef;
import io.gatling.javaapi.core.internal.Expressions;
import scala.Option;
import uk.co.developmentanddinosaurs.gatling.lambda.javaapi.action.InvokeActionBuilder;
import uk.co.developmentanddinosaurs.gatling.lambda.javaapi.protocol.LambdaProtocolBuilder;

public class LambdaDsl {

    public static final LambdaProtocolBuilder lambda =
            new LambdaProtocolBuilder(uk.co.developmentanddinosaurs.gatling.lambda.protocol.LambdaProtocolBuilder.apply(Predef.configuration()));

    private LambdaDsl(){}

    public static InvokeActionBuilder lambda(String functionName) {
        return new InvokeActionBuilder(uk.co.developmentanddinosaurs.gatling.lambda.action.InvokeActionBuilder.apply(Expressions.toStringExpression(functionName), Option.empty(), Option.empty()));
    }
}
